### PR TITLE
DPoP support 1st phase

### LINKED ISSUE #21200: DPoP support 1st phase
### Description

This issue is for supporting [KEYCLOAK-15169 OAuth 2.0 Demonstrating Proof-of-Possession at the Application Layer (DPoP)](https://github.com/keycloak/keycloak/pull/8895) partially.

 [KEYCLOAK-15169 OAuth 2.0 Demonstrating Proof-of-Possession at the Application Layer (DPoP)](https://github.com/keycloak/keycloak/pull/8895) can be broken down onto second phases.

1. Support DPoP not considering UserInfo endpoint
2. Support DPoP considering UserInfo endpoint

DPoP treatment for UserInfo endpoint have some issue for keycloak, therefore this point would be treated on the seconde phase.

This issue is to support the first part.

The difference between [the original PR](https://github.com/keycloak/keycloak/pull/8895) is as follows.
- not supported client configurations that are not defined by DPoP specification
- treated as PREVIEW feature
- added arquillian integration tests

Thanks to @dteleguin , the author of [the original PR](https://github.com/keycloak/keycloak/pull/8895). The PR for this issue is based on and includes [the original PR](https://github.com/keycloak/keycloak/pull/8895), and append some additional codes.

### Discussion

_No response_

### Motivation

_No response_

### Details

_No response_

### PR DESCRIPTION
**Pull request by tnorimat**
closes #21200 

This PR for supporting [KEYCLOAK-15169 OAuth 2.0 Demonstrating Proof-of-Possession at the Application Layer (DPoP)](https://github.com/keycloak/keycloak/pull/8895) partially.

 [KEYCLOAK-15169 OAuth 2.0 Demonstrating Proof-of-Possession at the Application Layer (DPoP)](https://github.com/keycloak/keycloak/pull/8895) can be broken down onto second phases.

Thanks to @dteleguin , the author of [the original PR](https://github.com/keycloak/keycloak/pull/8895). The PR for this issue is based on and includes [the original PR](https://github.com/keycloak/keycloak/pull/8895), and append some additional codes.

### CODE REVIEW

File: js/apps/admin-ui/src/clients/advanced/AdvancedSettings.tsx [hunk 1]
```diff
@@ -160,6 +160,35 @@ export const AdvancedSettings = ({
               )}
             />
           </FormGroup>
+          <FormGroup
```
Comment (jonkoops) [inline js/apps/admin-ui/src/clients/advanced/AdvancedSettings.tsx:163]: Should we be hiding this field based on whether the `DPOP` profile is enabled? We have this information in the environment variable injected to the front-end.

Comment (tnorimat) [reply js/apps/admin-ui/src/clients/advanced/AdvancedSettings.tsx:163]: Yes, we should hide the field about DPoP if DPoP profile is disabled. I will investigate into how to do it.

**Comment by tnorimat** [comment]
It seems that the fail of CodeQL is not relevant to the PR, not using unsecure hash algorithm (using SHA-256).

File: js/apps/admin-ui/src/clients/advanced/AdvancedSettings.tsx [hunk 2]
```diff
@@ -160,6 +165,37 @@ export const AdvancedSettings = ({
...
+                defaultValue={false}
+                control={control}
+                render={({ field }) => (
+                  <Switch
+                    id="oAuthDPoP-switch"
+                    label={t("common:on")}
+                    labelOff={t("common:off")}
+                    isChecked={field.value === "true"}
+                    onChange={(value) => field.onChange("" + value)}
+                    aria-label={t("oAuthDPoP")}
```
Comment (jonkoops) [inline js/apps/admin-ui/src/clients/advanced/AdvancedSettings.tsx:193]: If the `id` here matches the `fieldId`, I believe we do not need to add an `aria-label` attribute.

**Review by jonkoops — CHANGES_REQUESTED** [review]

File: js/apps/admin-ui/src/clients/advanced/AdvancedSettings.tsx (hunk 2 above)
Comment (tnorimat) [reply js/apps/admin-ui/src/clients/advanced/AdvancedSettings.tsx:193]: Thank you. I will fix it as you suggested.

**Comment by mposolda** [comment]
@dteleguin @tnorimat Thanks a lot for this effort! The PR looks quite good to me. I've did some review and testing and ended with sending follow-up PR https://github.com/Hitachi/keycloak/pull/203 to your branch with some small adjustements. What do you think about it? If you agree with it, feel free to merge to your branch, which should automatically update this PR.

Besides that, I can see following as possible follow-up work. However I think that these topics can be covered in the follow-up pull requests. What do you think? So my list is:
1. Documentation
2. Client registration support: There is the boolean switch `dpop_bound_access_tokens` mentioned in https://datatracker.ietf.org/doc/html/draft-ietf-oauth-dpop-11#section-5.2. We don't support it in this PR for the OIDC client registration requests. So I think that follow-up where this is added might be good.

Also I have these additional topics, which I would like to discuss. I don't see them as blockers for this PR, but maybe something, which we can consider as a follow-up if you agree.

3. Token type and introspection endpoint: Currently accessToken with DPoP thumbprint still have type `Bearer` in this PR. IMO it will makes more sense if the token type is `DPoP`. This is not explicitly mentioned in the specification, however it is mentioned that:
- Whole AccessTokenResponse should have token-type `DPoP`, which we do.
- The requests to resource-server also use `Authorization: DPoP ...` instead of `Authorization: Bearer ...` 
In general, the `DPoP` specification prescribes to use type `DPoP` in those places, so it would makes sense to me to use the same type in the access-token itself.
There is also one related point in the specification related to introspection endpoint https://datatracker.ietf.org/doc/html/draft-ietf-oauth-dpop-11#section-6.2 :
```
If the token_type member is included in the introspection response, it MUST contain the value DPoP
```
It seems to me that we're currently not doing it. We have some tests in the DPoPTest class, which use introspection endpoint, but they use token_type `access_token`. Which is not very accurate. It looks that for `DPoP` tokens (and also for `DPoP` token type), it will have more sense to use type `DPoP`. Maybe we should introduce new provider like `DPoPTokenIntrospectionProvider` ?


4. How to treat `dpop_bound_access_tokens`. The specification mentions:
```
dpop_bound_access_tokens
    Boolean value specifying whether the client always uses DPoP for token requests. If omitted, the default value is false.

If true, the authorization server MUST reject token requests from this client that do not contain the DPoP header.
```
My understanding is that:
- If `dpop_bound_access_tokens` is true, then DPoP is mandatory and Keycloak should reject token-request without `DPoP` header. It seems we're already doing that correctly in this PR
- If `dpop_bound_access_tokens` is false, then DPoP is not mandatory per the specification. However my understanding is, that `DPoP` request should be still possible for this client. IMO it is just not mandatory for this client, but in case that `DPoP` header is still included in the token-request, Keycloak may still verify `DPoP` header if present and attach to token thumbprint. Do you see it same way?


5. More use of client policies to ideally fully decouple DPoP from core Keycloak classes `TokenEndpoint` and `TokenManager`. However this one point as most problematic and it may require ability to define default client policies for the realm (which we don't yet have)...

**Comment by tnorimat** [comment]
@mposolda Thank you for your PR. I have merge it. Also I agree with follow-up tasks you suggested. I would like to create the issues for the follow-up tasks.

File: js/apps/admin-ui/src/clients/advanced/AdvancedSettings.tsx [hunk 3]
```diff
@@ -160,6 +165,37 @@ export const AdvancedSettings = ({
...
+              labelIcon={
+                <HelpItem
+                  helpText={t("clients-help:oAuthDPoP")}
+                  fieldLabelId="clients:oAuthDPoP"
+                />
+              }
+            >
+              <Controller
+                name={convertAttributeNameToForm<FormFields>(
+                  "attributes.dpop.bound.access.tokens"
```
Comment (mposolda) [inline js/apps/admin-ui/src/clients/advanced/AdvancedSettings.tsx:182]: This looks like same issue like in your other PR (The `iss` PR). Maybe you can just insert `,` as suggested?

**Review by mposolda — COMMENTED** [review]
@tnorimat Thanks!

From my point of view, we are quite good with this PR. I've added one inline comment, which hopefully helps with the UI  Attach files by dragging & dropping, selecting or pasting them.
test similarly like it helped for the `iss` PR. Also there is one additional inline comment by @jonkoops , which still seems to be unresolved? Can you please update as Jon suggested?

File: js/apps/admin-ui/src/clients/advanced/AdvancedSettings.tsx (hunk 3 above)
Comment (tnorimat) [reply js/apps/admin-ui/src/clients/advanced/AdvancedSettings.tsx:182]: Thank you I will fix it.

File: js/apps/admin-ui/src/clients/advanced/AdvancedSettings.tsx (hunk 2 above)
Comment (jonkoops) [reply js/apps/admin-ui/src/clients/advanced/AdvancedSettings.tsx:193]: @tnorimat have you taken a look at this?

**Review by jonkoops — DISMISSED** [review]
LGTM

**Comment by mposolda** [comment]
@jonkoops Thanks!

@tnorimat The failure in `HelpCommandDistTest` seems to be related to your changes? Probably something was added to Keycloak main in the meantime. Hopefully after rebase, you can figure it and fix it.

**Comment by tnorimat** [comment]
@tnorimat I fixed conflicts and rebase the PR on the latest main branch. One error you mentioned still happened but 
it seems that this fail of CodeQL is not relevant to the PR because the PR does not use unsecure hash algorithm the fail pointed ( the PR uses SHA-256, not unsecure SHA-1).

**Comment by tnorimat** [comment]
@mposolda I fixed conflicts and rebase the PR on the latest main branch. One error you mentioned still happened but
it seems that this fail of CodeQL is not relevant to the PR because the PR does not use unsecure hash algorithm the fail pointed ( the PR uses SHA-256, not unsecure SHA-1).

**Comment by tnorimat** [comment]
@mposolda I will check the failure of `HelpCommandDistTest`. Maybe resolving conflicts causes the failure.

**Review by mposolda — APPROVED** [review]
@tnorimat Thanks!

**Comment by mposolda** [comment]
@dteleguin @tnorimat Congrats on having this finally in. Thanks for your patience and all the work on this! I will create follow-up tasks discussed above.

**Comment by mposolda** [comment]
@tnorimat @dteleguin I've created the epic https://github.com/keycloak/keycloak/issues/21916 with the follow-up tasks for DPoP. I hope that most of them would be much easier than the initial PR. Does it makes sense to you?

**Comment by tnorimat** [comment]
@mposolda  @dteleguin Thank you. I will continue working on the issue of the epic https://github.com/keycloak/keycloak/issues/21916

**Comment by mposolda** [comment]
@tnorimat Thanks!

**Pull request merged**